csrc/multidevice/utils.cpp

```diff
@@ -165,6 +165,13 @@ bool haveDifferentShardings(
       continue;
     }
 
+    if (p_id->isBroadcast()) {
```

Comment:
Did you `!build`? I attempted to make a similar change months aog and found it broke some other unit tests.

Reply:
https://github.com/NVIDIA/Fuser/pull/3053 is the PR I was talking about. Unfortunately, it's too old to keep the test history; otherwise, I could figure out which test(s) failed.

Reply:
Hmm it is failing the single device multidevice tests... unfortunately there's two issues (1) size 1 axes are treated as broadcast regardless if it's the fake DID axis or an actual broadcast. (2) Reshapes that insert 'D' axes intended to be sharded and/or reduced are optimized away when D = 1.

Reply:
(1) isn't an issue for this PR, but it does come up with a sharded normalization op. (2) is unrelated to this particular line

Reply:
Removed this. Note, that (2) is still an issue in this test where an extra logical device axis is reshaped for a ReduceScatter that gets optimized away when D=1. 
Since this is an artifact of DID parallelizations only on the logical domain, I decided to skip this test when D=1 and left a TODO to revisit when DID loop parallelization is finished.

Reply:
> Reshapes that insert 'D' axes intended to be sharded and/or reduced are optimized away when D = 1.

Can you clarify the problem? IIUC, such a reshape will be converted into a broadcast. Similarly, a reshape that deletes `D` will be converted into a squeeze when `D=1`. In either case, the reshape isn't treated as resharding.

Reply:
The output of the local matmul (partially reduced):`[DIDx(D), i0, i1]`. In preparation for reduce scatter where i0 becomes the new sharded axis we have to reshape into `[DIDx(D), D, i0/D, i1]`. After this reshape the non-sharded `D` was optimized away. I tried turning off preseg passes remove bcast squeeze but a different error occurred.